### PR TITLE
 Subaru: Global gen1 longitudinal safety

### PR DESCRIPTION
**Pull request by jnewb1**
- OpenPilot/DBC refs needed updating because of AEB_Status being added
- Cleaned up subaru tests to use a base test that is extended by stock and op long tests
- Made macros in safety_subaru for common messages between gen1/gen2
- Added a common test for Gas/Brake cars (cars that don't have direct acceleration requests)
- subaru_longitudinal is behind the allow_debug flag

### CODE REVIEW

File: board/safety/safety_subaru.h [hunk 1]
```diff
@@ -20,6 +20,14 @@ const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = {
   .type = TorqueDriverLimited,
 };
 
+const LongitudinalLimits SUBARU_LONG_LIMITS = {
+  .min_gas = 0,
+  .max_gas = 3400,
+  .min_rpm = 0,
+  .max_rpm = 3200,
```
Comment (sshane) [inline board/safety/safety_subaru.h:27]: is RPM safety related? can setting it high cause more acceleration (gas) than we intend? If not, then we don't need to enforce it

File: board/safety_declarations.h [hunk 2]
```diff
@@ -220,6 +223,9 @@ uint32_t ts_angle_last = 0;
 int desired_angle_last = 0;
 struct sample_t angle_meas;         // last 6 steer angles/curvatures
 
+// for safety modes that need to passthru aeb
+bool stock_aeb = false;
+
```
Comment (sshane) [inline board/safety_declarations.h:228]: I think we can do AEB safety and passthrough in a separate PR. We make no guarantees for any alpha longitudinal behind the debug flag

File: tests/libpanda/safety_helpers.h [hunk 3]
```diff
@@ -168,6 +168,14 @@ void set_honda_fwd_brake(bool c){
...
 }
 
+void set_stock_aeb(bool c){
+  stock_aeb = c;
+}
+
+bool get_stock_aeb(){
+  return stock_aeb;
+}
+
```
Comment (sshane) [inline tests/libpanda/safety_helpers.h:178]: We might be able to combine honda's fwd_brake with a generic variable

File: board/safety/safety_subaru.h (hunk 1 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h:27]: Stock seems to scale throttle and rpm together, so I think its reasonable to limit it to the bounds of the stock system

File: board/safety_declarations.h (hunk 2 above)
Comment (jnewb1) [reply board/safety_declarations.h:228]: I'll remove aeb stuff for now and move it to a new PR

File: tests/libpanda/safety_helpers.h (hunk 3 above)
Comment (jnewb1) [reply tests/libpanda/safety_helpers.h:178]: Will make new PR for AEB stuff

File: board/safety/safety_subaru.h (hunk 1 above)
Comment (sshane) [reply board/safety/safety_subaru.h:27]: In that case, we can probably use one max value for both signals. Are they relatively close and move together?

Comment (jnewb1) [reply board/safety/safety_subaru.h:27]: rpm is typically a lot less, and will slowly ramp down even when throttle goes to zero

File: board/safety/safety_subaru.h [hunk 4]
```diff
@@ -20,6 +20,14 @@ const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = {
   .type = TorqueDriverLimited,
 };
 
+const LongitudinalLimits SUBARU_LONG_LIMITS = {
+  .min_gas = 0,
+  .max_gas = 3400,
```
Comment (sshane) [inline board/safety/safety_subaru.h:25]: Make sure to label how much acceleration in m/s^2 this applies

File: board/safety/safety_subaru.h [hunk 5]
```diff
@@ -20,6 +20,14 @@ const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = {
...
+  .min_rpm = 0,
+  .max_rpm = 3200,
+  .max_brake = 400,
```
Comment (sshane) [inline board/safety/safety_subaru.h:28]: Same, is this -4 m/s^2?

File: board/safety/safety_subaru.h (hunk 1 above)
Comment (sshane) [reply board/safety/safety_subaru.h:27]: Do you have any stock routes showing how the EyeSight system uses it? I would understand how it uses it and how it affects acceleration. If setting both to max values results in the desired max accel, we can limit both to max_gas

File: board/safety/safety_subaru.h (hunk 5 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h:28]: actual brake pressure. I will do some plots to see how brake and throttle relates to acceleration

Comment (jnewb1) [reply board/safety/safety_subaru.h:26]: brake is pretty linear. 400 is about -2.5m/s^2 according to c3 accelerometer 
image

File: board/safety/safety_subaru.h (hunk 1 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: image

RPM and Throttle are roughly related. I think just scaling them together will be sufficient for the initial PR, and I'll limit RPM to the same as throttle to make it easier.

File: board/safety/safety_subaru.h (hunk 4 above)
Comment (sshane) [reply board/safety/safety_subaru.h]: looks like it changes with speed, you can do something similar to GM like in https://github.com/commaai/openpilot/pull/25453 when you start to tune. Doesn't look like it can engine brake which is interesting. Maybe RPM can do that.

Comment (sshane) [reply board/safety/safety_subaru.h]: We usually allow 2 m/s in the safety, you can bump it

File: board/safety/safety_subaru.h [hunk 6]
```diff
@@ -20,6 +20,14 @@ const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = {
...
 };
 
+const LongitudinalLimits SUBARU_LONG_LIMITS = {
+  .max_brake = 400,  // approx -2.5 m/s^2
+
+  .min_gas = 0,
+  .max_gas = 3400,   // approx  1.2 m/s^2 at low speeds, less at higher speeds
+  .inactive_gas = 808,
+};
+
```
Comment (sshane) [inline board/safety/safety_subaru.h:30]: We usually allow 2 m/s^2 and -3.5 m/s^2 in safety

Comment (sshane) [inline board/safety/safety_subaru.h:30]:
```suggestion
const LongitudinalLimits SUBARU_LONG_LIMITS = {
  .max_gas = 3400,   // approx  1.2 m/s^2 at low speeds, less at higher speeds
  .inactive_gas = 808,
  .max_brake = 400,  // approx -2.5 m/s^2
};

```

corrected order. min_gas is only used if the range is not 0 to +x

Comment (jnewb1) [reply board/safety/safety_subaru.h]: I'll increase brake limit, but I'm going to leave gas limit as is until I can test max RPM and max gas in our subaru

File: tests/safety/common.py [hunk 7]
```diff
@@ -175,6 +174,53 @@ def test_accel_actuation_limits(self, stock_longitudinal=False):
           self.assertEqual(should_tx, self._tx(self._accel_msg(accel)))
 
 
+class LongitudinalGasBrakeSafetyTest(PandaSafetyTestBase, abc.ABC):
```
Comment (sshane) [inline tests/safety/common.py:177]: Nice, I wanted to eventually do this, but I forgot if there was a snag making it generic. Feel free to do this in another PR before or after this one

File: tests/safety/common.py [hunk 8]
```diff
@@ -175,6 +174,53 @@ def test_accel_actuation_limits(self, stock_longitudinal=False):
...
+  MIN_BRAKE: int = 0
+  MAX_BRAKE: Optional[int] = None
+
+  MIN_THROTTLE: int = 1818
```
Comment (sshane) [inline tests/safety/common.py:181]: let's use gas for now, it's what the other brands use

File: tests/safety/common.py (hunk 7 above)
Comment (jnewb1) [reply tests/safety/common.py:177]: once i get this merged I'll look at cleaning up the safety tests for the other brands too

File: tests/safety/common.py (hunk 8 above)
Comment (jnewb1) [reply tests/safety/common.py]: done

**Comment by jnewb1** [comment]
Got a safety replay from martin, this should be good to merge now @sshane

**Comment by sshane** [comment]
Great, I'll look over everything soon!

File: board/safety/safety_subaru.h [hunk 9]
```diff
@@ -109,18 +130,18 @@ static int subaru_rx_hook(CANPacket_t *to_push) {
...
     // update vehicle moving with any non-zero wheel speed
-    if ((addr == MSG_SUBARU_Wheel_Speeds) && (bus == alt_bus)) {
+    if ((addr == MSG_SUBARU_Wheel_Speeds) && (bus == alt_main_bus)) {
       vehicle_moving = ((GET_BYTES(to_push, 0, 4) >> 12) != 0U) || (GET_BYTES(to_push, 4, 4) != 0U);
     }
 
-    if ((addr == MSG_SUBARU_Brake_Status) && (bus == alt_bus)) {
-      brake_pressed = ((GET_BYTE(to_push, 7) >> 6) & 1U);
+    if ((addr == MSG_SUBARU_Brake_Status) && (bus == alt_main_bus)) {
+      brake_pressed = GET_BIT(to_push, 62U) != 0U;
```
Comment (sshane) [inline board/safety/safety_subaru.h:144]: you can do the variable rename and these GET_BIT refactors in another PR if you want

File: board/safety/safety_subaru.h [hunk 10]
```diff
@@ -161,7 +203,11 @@ static int subaru_fwd_hook(int bus_num, int addr) {
   int bus_fwd = -1;
 
   if (bus_num == SUBARU_MAIN_BUS) {
-    bus_fwd = SUBARU_CAM_BUS;  // forward to camera
+    // Global Platform
+    bool block_msg = subaru_longitudinal && ((addr == MSG_SUBARU_Brake_Status) || (addr == MSG_SUBARU_CruiseControl));
+    if (!block_msg) {
+      bus_fwd = SUBARU_CAM_BUS;  // to the eyesight camera
+    }
```
Comment (sshane) [inline board/safety/safety_subaru.h:210]: What does this do? Why block these messages to the camera? Comment explaining this would be a good idea

**Review by sshane — CHANGES_REQUESTED** [review]

File: board/safety/safety_subaru.h [hunk 11]
```diff
@@ -14,6 +14,12 @@ const SteeringLimits SUBARU_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(2
 const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(1000, 40, 40);
 
 
+const LongitudinalLimits SUBARU_LONG_LIMITS = {
+  .max_gas = 3400,      // approx  1.2 m/s^2 at low speeds, less at higher speeds
+  .inactive_gas = 1818, // this is zero acceleration
```
[the remark this answers is not in the document]
Comment (jnewb1) [reply board/safety/safety_subaru.h:19]: 1818 is zero acceleration, 3400 is full, and 808 appears to be engine braking (set to 808 when braking value is high)

[the remark this answers is not in the document]
Comment (sshane) [reply board/safety/safety_subaru.h:19]: then make sure to set min_gas and clip to 808 in openpilot. check GM, it's very similar (there's a zero accel and regen gas value)

[the remark this answers is not in the document]
Comment (sshane) [reply board/safety/safety_subaru.h:19]: since 3400 is only 1.2 m/s^2, does the stock system ever go above? maybe the RPM adds on to the acceleration somewhat

File: board/safety/safety_subaru.h [hunk 12]
```diff
@@ -14,6 +14,12 @@ const SteeringLimits SUBARU_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(2
 const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(1000, 40, 40);
 
 
+const LongitudinalLimits SUBARU_LONG_LIMITS = {
+  .max_gas = 3400,      // approx  1.2 m/s^2 at low speeds, less at higher speeds
```
[the remark this answers is not in the document]
Comment (jnewb1) [reply board/safety/safety_subaru.h:18]: it goes above 4000 when on the highway and accelerating

File: board/safety/safety_subaru.h (hunk 10 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: Brake_Status and CruiseControl are intercepted and rewritten in openpilot.

Brake_Status is to fake the ES_Brake feedback signal to eyesight, which was for preserving AEB.
CruiseControl is to tell eyesight that it is not enabled (we set Cruise_Activated to false), so it only will try and perform AEB

File: board/safety/safety_subaru.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (jnewb1) [reply board/safety/safety_subaru.h]: I'll do some testing with RPM and Throttle on our outback. Stock goes up to 4000 when accelerating on the highway

File: board/safety/safety_subaru.h [hunk 13]
```diff
@@ -132,13 +153,17 @@ static int subaru_rx_hook(CANPacket_t *to_push) {
   return valid;
 }
 
+
 static int subaru_tx_hook(CANPacket_t *to_send) {
 
   int tx = 1;
   int addr = GET_ADDR(to_send);
+  bool violation = false;
```
[the remark this answers is not in the document]
Comment (jnewb1) [reply board/safety/safety_subaru.h:160]: refactored to use "violation" everywhere

File: board/safety/safety_subaru.h (hunk 10 above)
Comment (sshane) [reply board/safety/safety_subaru.h]: Why can't you let the EyeSight camera see the car's signals? Will it thinking it's in control of ACC somehow change the AEB behavior?

@adeebshihadeh told me that when working on the first iteration of an experimental longitudinal support PR, we don't need or want to take the time to verify that AEB still works, since we don't guarantee that, and when we do ship this to release (not master), we will need to re-do AEB testing anyway.

If there are no visible faults, I think we should remove this blocking/forwarding (complexity) and just show the AEB disabled light for now.

File: tests/safety/common.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jnewb1) [reply tests/safety/common.py:178]: New PR for making this test. The only other brand I saw that was using gas/brake was GM, do you know of any others?

https://github.com/commaai/panda/pull/1537

[the remark this answers is not in the document]
Comment (sshane) [reply tests/safety/common.py:178]: Think honda might? Also Ford. Though it's a bit different, they implement a gas check and accel check (for braking only). Might want classes to individually test each, or a class for these two cases as it seems common.

[the remark this answers is not in the document]
Comment (sshane) [reply tests/safety/common.py:178]: Though since we don't have a nice way to know the max ranges of signals, you can keep the gas/brake tests locally to Subaru for now, and we can clean up in a future PR.

File: board/safety/safety_subaru.h [hunk 14]
```diff
@@ -170,7 +216,13 @@ static int subaru_fwd_hook(int bus_num, int addr) {
                        (addr == MSG_SUBARU_ES_DashStatus) ||
                        (addr == MSG_SUBARU_ES_LKAS_State) ||
                        (addr == MSG_SUBARU_ES_Infotainment));
-    if (!block_lkas) {
+
+    bool block_long = ((addr == MSG_SUBARU_ES_Brake) ||
+                       (addr == MSG_SUBARU_ES_Distance) ||
+                       (addr == MSG_SUBARU_ES_Status));
```
[the remark this answers is not in the document]
Comment (jnewb1) [reply board/safety/safety_subaru.h:214]: c3d1ccb52f5f9d65|2023-08-01--15-02-58 -> sending max cruise_rpm but very low cruise_throttle (just above inactive), car very slowly accelerates, and the engine eventually gets to cruise_rpm

[the remark this answers is not in the document]
Comment (sshane) [reply board/safety/safety_subaru.h:214]: so does the acceleration correspond to cruise_throttle, or is it slightly higher? and does sending both high cause more acceleration than just cruise_throttle or what you were doing in openpilot (can the RPM allow more acceleration than what you've defined in safety)?

[the remark this answers is not in the document]
Comment (jnewb1) [reply board/safety/safety_subaru.h:214]: c3d1ccb52f5f9d65|2023-08-01--16-16-02 -> maxing cruise_throttle (3400) and cruise_rpm (3200) gives around 3 m/s^2 from a stop
c3d1ccb52f5f9d65|2023-08-01--16-26-58 -> maxing cruise_throttle (3400) and a low cruise_rpm (1200) gives more like 2 m/s^2 from stop

So it seems like we should also limit cruise_rpm as it does have some effect on acceleration

**Comment by sshane** [comment]
It looks like ES_Status->Cruise_RPM is just a request of the transmission RPM, can we just set it to some median value at all times(since it seems related to the speed), and use the throttle, or does that reduce some performance?

image

File: board/safety/safety_subaru.h (hunk 9 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h:143]: Moved these small changes to new PR: https://github.com/commaai/panda/pull/1543

File: board/safety/safety_subaru.h [hunk 15]
```diff
@@ -14,6 +14,13 @@ const SteeringLimits SUBARU_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(2
 const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(1000, 40, 40);
 
 
+const LongitudinalLimits SUBARU_LONG_LIMITS = {
+  .min_gas = 808,
+  .max_gas = 3400,      // approx  2 m/s^2 when maxing cruise_rpm and cruise_throttle
+  .inactive_gas = 1818, // this is zero acceleration, and 808 appears to be engine braking
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:20]:
```suggestion
  .inactive_gas = 1818, // this is zero acceleration
```

File: board/safety/safety_subaru.h [hunk 16]
```diff
@@ -195,6 +243,10 @@ static int subaru_fwd_hook(int bus_num, int addr) {
 static const addr_checks* subaru_init(uint16_t param) {
   subaru_gen2 = GET_FLAG(param, SUBARU_PARAM_GEN2);
 
+#ifdef ALLOW_DEBUG
+  subaru_longitudinal = GET_FLAG(param, SUBARU_PARAM_LONGITUDINAL);
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:247]: also gate on not gen2?

File: board/safety/safety_subaru.h [hunk 17]
```diff
@@ -165,7 +187,27 @@ static int subaru_tx_hook(CANPacket_t *to_send) {
     violation |= steer_torque_cmd_checks(desired_torque, -1, limits);
   }
 
-  if (violation){
+  if (subaru_longitudinal) {
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:190]: this shouldn't be needed since this only further blacklists messages. if we're not doing long, tx should already be false from the `msg_allowed` check

File: board/safety/safety_subaru.h [hunk 18]
```diff
@@ -165,7 +187,27 @@ static int subaru_tx_hook(CANPacket_t *to_send) {
...
+
+    // check es_distance cruise_throttle limits
+    if (addr == MSG_SUBARU_ES_Distance) {
+      int cruise_throttle = ((GET_BYTES(to_send, 0, 4) >> 16) & 0xFFFU);
+      violation |= longitudinal_gas_checks(cruise_throttle, SUBARU_LONG_LIMITS);
+    }
+
+    // check es_status cruise_rpm limits
+    if (addr == MSG_SUBARU_ES_Status) {
+      int cruise_rpm = ((GET_BYTES(to_send, 0, 4) >> 16) & 0xFFFU);
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:205]: simplify this, it's getting two extra bytes then immediately shifting them out

File: board/safety/safety_subaru.h [hunk 19]
```diff
@@ -165,7 +187,27 @@ static int subaru_tx_hook(CANPacket_t *to_send) {
     violation |= steer_torque_cmd_checks(desired_torque, -1, limits);
   }
 
-  if (violation){
+  if (subaru_longitudinal) {
+    // check es_brake brake_pressure limits
+    if (addr == MSG_SUBARU_ES_Brake) {
+      int es_brake_pressure = ((GET_BYTES(to_send, 0, 4) >> 16) & 0xFFFFU);
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:193]: same here

File: board/safety/safety_subaru.h [hunk 20]
```diff
@@ -165,7 +187,27 @@ static int subaru_tx_hook(CANPacket_t *to_send) {
...
+      violation |= longitudinal_brake_checks(es_brake_pressure, SUBARU_LONG_LIMITS);
+    }
+
+    // check es_distance cruise_throttle limits
+    if (addr == MSG_SUBARU_ES_Distance) {
+      int cruise_throttle = ((GET_BYTES(to_send, 0, 4) >> 16) & 0xFFFU);
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:199]: and here

**Comment by adeebshihadeh** [comment]
Also rebase on master to get the HITL test fix

File: board/safety/safety_subaru.h (hunk 20 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: done

File: board/safety/safety_subaru.h (hunk 19 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: done

File: board/safety/safety_subaru.h (hunk 18 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: done

File: board/safety/safety_subaru.h (hunk 17 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: done

File: board/safety/safety_subaru.h (hunk 16 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: done

File: board/safety/safety_subaru.h (hunk 15 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: moved it up by the min_gas field

File: tests/safety/test_subaru.py [hunk 21]
```diff
@@ -101,6 +104,52 @@ class TestSubaruGen2SafetyBase(TestSubaruSafetyBase):
...
+
+  MIN_BRAKE = 0
+  MAX_BRAKE = 600
+  MAX_POSSIBLE_BRAKE = 2**16
+
+  FWD_BLACKLISTED_ADDRS = {2: [MSG_SUBARU_ES_LKAS, MSG_SUBARU_ES_Brake, MSG_SUBARU_ES_Distance,
+                               MSG_SUBARU_ES_Status, MSG_SUBARU_ES_DashStatus,
+                               MSG_SUBARU_ES_LKAS_State, MSG_SUBARU_ES_Infotainment]}
+
+  def _generic_limit_safety_check(self, msg_function, min_value, max_value, max_possible_value, inactive_value=0):
```
Comment (adeebshihadeh) [inline tests/safety/test_subaru.py:122]: doesn't seem subaru specific. @sshane do we have something like this somewhere else?

Comment (jnewb1) [reply tests/safety/test_subaru.py:122]: I started one here: https://github.com/commaai/panda/pull/1537

RPM would be specific to subaru, and some cars use an acceleration request for brake, and throttle-like message for gas. But I can look at cleaning it up into a common test

Comment (adeebshihadeh) [reply tests/safety/test_subaru.py:122]: yeah let's do that first

File: board/safety/safety_subaru.h (hunk 17 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: I think we do want this check actually, because we use ES_Distance for a cancel request even when stock long is enabled, and it might get blocked here if eyesight happens to be doing the max accel (I've seen it go to 4000)

Comment (adeebshihadeh) [reply board/safety/safety_subaru.h]: that shouldn't be allowed in stock long. can you make a pair of PRs to openpilot and panda to fix that? no actuation should be allowed while cancelling

File: tests/safety/test_subaru.py (hunk 21 above)
Comment (jnewb1) [reply tests/safety/test_subaru.py]: done https://github.com/commaai/panda/pull/1537

File: board/safety/safety_subaru.h (hunk 17 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: Done

File: tests/safety/test_subaru.py [hunk 22]
```diff
@@ -100,6 +103,40 @@ def test_cancel_message(self):
...
+  INACTIVE_GAS = 1818
+  MAX_POSSIBLE_GAS = 2**12
+
+  MIN_BRAKE = 0
+  MAX_BRAKE = 600
+  MAX_POSSIBLE_BRAKE = 2**16
+
+  MIN_RPM = 0
+  MAX_RPM = 2400
+  MAX_POSSIBLE_RPM = 2**12
```
Comment (adeebshihadeh) [inline tests/safety/test_subaru.py:118]: ideally we get this from the DBC

File: board/safety/safety_subaru.h [hunk 23]
```diff
@@ -166,12 +191,32 @@ static int subaru_tx_hook(CANPacket_t *to_send) {
...
+  if (addr == MSG_SUBARU_ES_Distance) {
     int cruise_throttle = (GET_BYTES(to_send, 2, 2) & 0xFFFU);
     bool cruise_cancel = GET_BIT(to_send, 56U) != 0U;
-    violation |= (cruise_throttle != 1818);
-    violation |= (!cruise_cancel);
+    
+    if (subaru_longitudinal) {
+      violation |= longitudinal_gas_checks(cruise_throttle, SUBARU_LONG_LIMITS);
+    }
+    else {
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:208]: same line

File: board/safety/safety_subaru.h [hunk 24]
```diff
@@ -14,6 +14,16 @@ const SteeringLimits SUBARU_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(2
 const SteeringLimits SUBARU_GEN2_STEERING_LIMITS = SUBARU_STEERING_LIMITS_GENERATOR(1000, 40, 40);
 
 
+const LongitudinalLimits SUBARU_LONG_LIMITS = {
+  .min_gas = 808,       // appears to be engine braking
+  .max_gas = 3400,      // approx  2 m/s^2 when maxing cruise_rpm and cruise_throttle
+  .inactive_gas = 1818, // this is zero acceleration
+  .max_brake = 600,     // approx -3.5 m/s^2
+
+  .min_alt = 0,         // Transmission RPM limits
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:23]: this alt name is confusing

Comment (jnewb1) [reply board/safety/safety_subaru.h:23]: the goal was to make it generic in case any other brands have a similar "alternative" gas or brake signal that needs limiting. "rpm" is also weird, because it's actually the transmission rpm not the engine rpm

Comment (adeebshihadeh) [reply board/safety/safety_subaru.h:23]: it's **incredibly important** that this particular code is as simple and readable as possible. in this case, the variable should probably be called `min_transmission_rpm`

File: board/safety/safety_subaru.h (hunk 23 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: done

File: board/safety/safety_subaru.h (hunk 24 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: renamed to transmission_rpm

File: tests/safety/test_subaru.py (hunk 22 above)
Comment (sshane) [reply tests/safety/test_subaru.py:118]: yeah we should add to CANDefine (right now it just gives you the value map, but we can expand to give general signal info)
